sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,774 @@
+// the cache until we actually re-store it.
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decompress packed storage entry: {{err}}", err)
+	}
+	if notCompressed {
+		uncompressedData = storageEntry.Value
+	}
+
+	var bucket Bucket
+	err = proto.Unmarshal(uncompressedData, &bucket)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode packed storage entry: {{err}}", err)
+	}
+
+	cacheKey := s.GetCacheKey(storageEntry.Key)
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lb := &LockedBucket{
+		LockEntry: lock,
+		Bucket:    &bucket,
+	}
+	lb.Key = storageEntry.Key
+
+	return lb, nil
+}
+
+// Put stores a packed bucket entry
+func (s *StoragePackerV2) PutBucket(ctx context.Context, bucket *LockedBucket) error {
+	if bucket == nil {
+		return fmt.Errorf("nil bucket entry")
+	}
+
+	if bucket.Key == "" {
+		return fmt.Errorf("missing key")
+	}
+
+	bucket.Lock()
+	defer bucket.Unlock()
+
+	if err := s.storeBucket(ctx, bucket, true); err != nil {
+		return errwrap.Wrapf("failed at high level bucket put: {{err}}", err)
+	}
+
+	s.bucketsCacheLock.Lock()
+	s.bucketsCache.Insert(s.GetCacheKey(bucket.Key), bucket)
+	s.bucketsCacheLock.Unlock()
+
+	return nil
+}
+
+func (s *StoragePackerV2) shardBucket(ctx context.Context, bucket *LockedBucket, cacheKey string, allowLocking bool) error {
+	if allowLocking {
+		s.shardLock.Lock()
+		defer s.shardLock.Unlock()
+
+		bucketLock := bucket.LockEntry
+		defer func() {
+			for lock := range shardLocks {
+				// Let the initial calling function take care of the highest level lock
+				if lock != bucketLock {
+					lock.Unlock()
+				}
+			}
+
+			// Empty the map
+			shardLocks = make(map[*locksutil.LockEntry]struct{}, 32)
+		}()
+	}
+
+	numShards := int(math.Pow(2.0, float64(s.BucketShardBits)))
+
+	// Create the shards and lock them
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shardLocks[bucket.LockEntry] = struct{}{}
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
+		shardedBucket := &LockedBucket{
+			LockEntry: lock,
+			Bucket: &Bucket{
+				Key:     fmt.Sprintf("%s/%s", bucket.Key, shardKey),
+				ItemMap: make(map[string][]byte),
+			},
+		}
+		shards[shardKey] = shardedBucket
+		// If it was equal we'd be locked already
+		s.Logger.Trace("created shard", "shard_key", shardKey)
+		// Don't try to lock the same lock twice in case it hashes that way
+		if _, ok := shardLocks[lock]; !ok {
+			s.Logger.Trace("locking lock", "shard_key", shardKey)
+			lock.Lock()
```

Comment:
This doesn't seem deadlock-safe to me. The locks from s.storageLocks can be acquired in some arbitrary order, correct? So any guarantee depending on lock ordering goes out the window. Maybe it's safe if no non-sharding thread acquires more than one bucket lock, and we do have the lock to ensure only one sharing operation at a time.

However, I am worried about the following sequence:
A acquires storageLock 1 in PutBucket
B acquired storageLock 2 in PutBucket for a different bucket
A needs to shard, acquires shardLock
B also needs to shard, blocks on shardLock
A's cacheKey+shardKey duplicates to storageLock 2, deadlock?

I think it is a problem to acquire storageLocks both before and after shardLock.

Reply:
I think you're right. I'm not sure how to best solve this -- the locking is pretty nightmarish already.

It's possible that the answer is that we just use one lock. SP v1 used sharded locks but didn't have a read cache -- we do here. So we only need to write lock when actually writing a bucket (or read lock it when there's a cache miss). 

Thoughts?

Reply:
Hm, I thought the go race detector also did deadlock detection, but I guess not. I wrote a small stress test which exhibited deadlock. Looks like the one I identified, goroutine 10 is trying to perform s.shardLock.Lock:

```
goroutine 10 [semacquire, 3 minutes]:
sync.runtime_SemacquireMutex(0xc0001b6078, 0x900000000)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0001b6074)
	/usr/local/go/src/sync/mutex.go:134 +0x1e2
sync.(*RWMutex).Lock(0xc0001b6074)
	/usr/local/go/src/sync/rwmutex.go:93 +0x4a
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).shardBucket(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc00016eb20, 0xc0001789a3, 0x3, 0x1, 0x0, 0x0)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:306 +0x2356
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).storeBucket(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc00016eb20, 0x1, 0xd4ab80, 0xc00016eb20)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:450 +0x42a
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).PutItem(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc000020460, 0x0, 0x0)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:696 +0x56a
github.com/hashicorp/vault/sdk/helper/storagepacker.TestStoragePackerV2_Race.func1(0xc0000c8580, 0xc8, 0x190)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2_test.go:478 +0x10b
created by github.com/hashicorp/vault/sdk/helper/storagepacker.TestStoragePackerV2_Race
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2_test.go:485 +0x605

goroutine 11 [semacquire, 3 minutes]:
sync.runtime_SemacquireMutex(0xc0004dc924, 0x900000000)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0004dc920)
	/usr/local/go/src/sync/mutex.go:134 +0x1e2
sync.(*RWMutex).Lock(0xc0004dc920)
	/usr/local/go/src/sync/rwmutex.go:93 +0x4a
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).shardBucket(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc000162d80, 0xc00017890b, 0x3, 0x1, 0x0, 0x0)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:348 +0xcac
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).storeBucket(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc000162d80, 0x1, 0xd4ab80, 0xc000162d80)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:450 +0x42a
github.com/hashicorp/vault/sdk/helper/storagepacker.(*StoragePackerV2).PutItem(0xc0001b6000, 0xe84f40, 0xc000174210, 0xc00027b9a0, 0x0, 0x0)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2.go:696 +0x56a
github.com/hashicorp/vault/sdk/helper/storagepacker.TestStoragePackerV2_Race.func1(0xc0000c8bc0, 0xc8, 0xc8)
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2_test.go:478 +0x10b
created by github.com/hashicorp/vault/sdk/helper/storagepacker.TestStoragePackerV2_Race
	/home/mgritter/vault/sdk/helper/storagepacker/storagepacker_v2_test.go:486 +0x644
```

I'll give this a bit more thought and propose a solution tomorrow.